src/sources/dnstap/parser.rs

```diff
@@ -0,0 +1,1105 @@
+            format!("{:?}", to_dnstap_message_type(dnstap_message_type_id)?),
+        );
+
+        let request_message_key = self
+            .event_schema
+            .dnstap_message_schema()
+            .request_message()
+            .to_string();
+        let response_message_key = self
+            .event_schema
+            .dnstap_message_schema()
+            .response_message()
+            .to_string();
+
+        if let Some(query_time_sec) = dnstap_message.query_time_sec {
+            let mut time_in_nanosec: i64 = query_time_sec as i64 * 1_000_000_000_i64;
+            let mut query_time_nsec = 0;
+            if let Some(nsec) = dnstap_message.query_time_nsec {
+                query_time_nsec = nsec;
+                time_in_nanosec += query_time_nsec as i64;
+            }
+
+            if DNSTAP_MESSAGE_REQUEST_TYPE_IDS.contains(&dnstap_message_type_id) {
+                self.log_time(
+                    self.event_schema.dnstap_root_data_schema().time(),
+                    time_in_nanosec,
+                    self.event_schema.dnstap_root_data_schema().time_precision(),
+                    "ns",
+                );
+
+                let timestamp = Utc.timestamp(query_time_sec.try_into().unwrap(), query_time_nsec);
+                self.insert(
+                    self.event_schema.dnstap_root_data_schema().timestamp(),
+                    timestamp,
+                );
+            }
+
+            if dnstap_message.query_message != None {
+                self.parent_key_path
+                    .push(PathComponent::Key(request_message_key.clone()));
+
+                let time_key_name = if dnstap_message_type_id <= 12 {
+                    self.event_schema.dns_query_message_schema().time()
+                } else {
+                    self.event_schema.dns_update_message_schema().time()
+                };
+
+                let time_precision_key_name = if dnstap_message_type_id <= 12 {
+                    self.event_schema
+                        .dns_query_message_schema()
+                        .time_precision()
+                } else {
+                    self.event_schema
+                        .dns_update_message_schema()
+                        .time_precision()
+                };
+
+                self.log_time(
+                    time_key_name,
+                    time_in_nanosec,
+                    time_precision_key_name,
+                    "ns",
+                );
+
+                self.parent_key_path.pop();
+            }
+        }
+
+        if let Some(response_time_sec) = dnstap_message.response_time_sec {
+            let mut time_in_nanosec: i64 = response_time_sec as i64 * 1_000_000_000_i64;
+
+            let mut response_time_nsec = 0;
+            if let Some(nsec) = dnstap_message.response_time_nsec {
+                response_time_nsec = nsec;
+                time_in_nanosec += response_time_nsec as i64;
+            }
```

Comment:
Nit: you could avoid needing a mutable variable here via:

```
let (time_in_nanosec, response_time_nsec) = match dnstap_message.response_time_nsec {
  Some(nsec) => (response_time_sec * 1_000_000_000_i64 + nsec, nsec),
  None => (response_time_sec * 1_000_000_000_i64, 0)
}
```

I usually find things easier to reason about if they are immutable unless necessary.

Also, why do we do this addition here if `response_time_nsec` is `Some`?

Reply:
Good suggestions. Will change the code to not use mutable variables here. Also same for `dnstap_message.query_time_nsec` above. Regarding the addition logic, we need to combine both response time in seconds and response time in nano seconds to calculate the total `time_in_nanosec` if `dnstap_message.response_time_nsec` is not None.

Reply:
Updated.